### PR TITLE
Added New parameter ProductCoveredbySA that is introduced in SQL 2022

### LINKED ISSUE #2040: New parameter productcoveredbysa : Command proposal
### Command proposal

A new installation parameter, /PRODUCTCOVEREDBYSA, was introduced in SQL Server 2022 (16.x). This parameter indicates whether the provided product key (/PID=) license is covered under a Software Assurance or SQL Server Subscription contract, or just a SQL Server license.

image

### Proposed parameters

Parameter | Mandatory | Data type | Description | Default value | Allowed values
--- | --- | --- | --- | --- | ---
productcoveredbysa | yes | String | Detailed description | None | true,False

"PRODUCTCOVEREDBYSA"

### Special considerations or limitations

We have the **new** parameter "PRODUCTCOVEREDBYSA" available for "Install-SqlDscResource" but currently there is no resource using that command and this is to add the parameter to the DSC resource "**DSC_SqlSetup**" for the SQL 2022 installs.

SqlServerDsc-16.6.0\source\DSCResources\DSC_SqlSetup

https://github.com/dsccommunity/SqlServerDsc/issues/1798

### PR DESCRIPTION
**Pull request by sara921-spec**
This Pull Request (PR) fixes the following issues

#### Pull Request (PR) description
- 'ProductCoveredbySA'
  - Added new parameter ProductCoveredbySA which is introduced in SQL 2022.

 Adding parameter ProductCoveredbySA (introduced in sql 2022) into DSC_SqlSetup
      Specifies the license coverage for SQL Server. True indicates it's covered under Software Assurance or SQL Server 
      subscription.
      False, or omitting the parameter, indicates it's covered under a SQL Server license.
      Default value is False.

#### This Pull Request (PR) fixes the following issues
- Fixes #2040

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [ ] Resource documentation updated in the resource's README.md.
- [x] Resource parameter descriptions updated in schema.mof.
- [x] Comment-based help updated, including parameter descriptions.
- [ ] Localization strings updated.
- [x] Examples updated.
- [x] Unit tests updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Integration tests updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Code changes adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

- - -
[Reviewable](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044)

### CODE REVIEW

**Review by johlju — CHANGES_REQUESTED** [review]
Thank you for sending this in.

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-:-O5JVO5G6UsQOzlRIPwh:boepzvk)* status: all files reviewed, 2 unresolved discussions (waiting on @sara921-spec)

*[a discussion](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O5JVJplA8kY4MIy0qSG:-O5JVJpmCCxJEUFiuIM0:b-45qglh) (no related file):*
Please also update the schema: https://github.com/dsccommunity/SqlServerDsc/blob/main/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.schema.mof

___
___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 1319 at r2](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O5JVm2WDitHLyUN1ooE:-O5JVm2WDitHLyUN1ooF:bvbxrbf) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/c198aac83232164e1625931e0a95df39f237504c/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L1319)):*
> ```PowerShell
>         {
>             $setupArguments['ProductCoveredBySA'] = $false
>         }
> ```

I think we can skip this? Omitting the parameter from the arguments is the same as saying `/PRODUCTCOVEREDBYSA=False`. It will also simplify the unit test cases needed.



_Code quote:_
```PowerShell
        else
        {
            $setupArguments['ProductCoveredBySA'] = $false
        }
```

**Review by johlju — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044)* status: all files reviewed, 3 unresolved discussions (waiting on @sara921-spec)

___
*[`CHANGELOG.md` line 11 at r2](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O5JXeYc8B3iQDHAxkWH:-O5JXeYc8B3iQDHAxkWI:b-dj742b) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/c198aac83232164e1625931e0a95df39f237504c/CHANGELOG.md#L11)):*
> ```Markdown
> 
> - 'ProductCoveredbySA'
>  - Added new parameter ProductCoveredbySA which in introduced in SQL 2022.
> ```

Suggest changing to the below. Use SqlSetup and make sure second entry is indented with 2 spaces for the list to render correctly.



_Suggestion:_
```Markdown
- SqlSetup
  - Added new parameter ProductCoveredbySA which in introduced in SQL 2022.
```

**Comment by sara921-spec** [comment]
Hi

**Comment by sara921-spec** [comment]
hi @johlju can u help me out on the unit test? on where I can possibly add the HKLM Path that it is failing on

Thanks!

**Comment by johlju** [comment]
Happy to take a look, but might take a few days before I have time. 😊

**Review by johlju — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044)* status: 1 of 4 files reviewed, 2 unresolved discussions (waiting on @sara921-spec)

___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 325 at r3](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O5PKdXP2B13VK8iIbhh:-O5PKdXQD5Udo7-O0llU:b-8yud3p) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/c3ffdb5bb2c83703f0fb2e7f4c6a97a755e47463/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L325)):*
> ```PowerShell
>         {
>            # Grab the value of ProductCoveredBySA from the registry based on the instance
>            $getTargetResourceReturnValue.ProductCoveredBySA = Get-ItemProperty "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL$($SqlVersion).$($InstanceName)\Setup\IsProductCoveredBySA"
> ```

Instead of Get-ItemProperty, try using the helper function `Get-RegistryPropertyValue`: https://github.com/dsccommunity/SqlServerDsc/blob/3c3092976409645d74f58707331f66ffe1967127/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1#L18C10-L18C35

But you probably have to test that it actually can get the value in a real environment using the helper function. :thinking: .



_Code quote:_
```PowerShell
Get-ItemProperty
```

**Comment by johlju** [comment]
If the above command (in the review comment) is able to get the value, then it will should resolve the error in the tests. If the command does not work, then you need to handle the error being thrown another way. 🙂

**Comment by sara921-spec** [comment]
> If the above command (in the review comment) is able to get the value, then it will should resolve the error in the tests. If the command does not work, then you need to handle the error being thrown another way. 🙂

@johlju This update to use "Get-RegistryPropertyValue" as suggested worked to pass the HKLM path errors, we have only 2 check failures now. 

error - 'Conn' is not a valid value for setting 'FEATURES'.

**Comment by johlju** [comment]
The Feature [CONN and some other features](https://learn.microsoft.com/en-us/sql/database-engine/install-windows/install-sql-server-from-the-command-prompt?view=sql-server-ver16#Feature) are not available in SQL Server 2022. I see you have enabled SQL Server 2022 (major version '16') for a bunch of tests which is good because that is really needed, but that is why it is failing. You need to make sure the correct values are used and then evaluated when the major version is 16. For example not calling the *-TargetResource with 'CONN' features parameter.

**Comment by sara921-spec** [comment]
> The Feature [CONN and some other features](https://learn.microsoft.com/en-us/sql/database-engine/install-windows/install-sql-server-from-the-command-prompt?view=sql-server-ver16#Feature) are not available in SQL Server 2022. I see you have enabled SQL Server 2022 (major version '16') for a bunch of tests which is good because that is really needed, but that is why it is failing. You need to make sure the correct values are passed to *-TargetResource and then evaluated the correct values where returned when the major version is 16. For example not calling the *-TargetResource with 'CONN' features parameter.

There is a "conn" here as in the below img, do u suggest adding a separate line for **version 16** without conn?

image

image

**Comment by sara921-spec** [comment]
Hi @johlju What are the available features for 2022 or how to get that? any idea?

https://github.com/dsccommunity/SqlServerDsc/issues/1566

Seems "RS" is not valid too

Error:  Context When installing a named instance for major version 16
   Context When installing all features
##[error]      [-] Should set the system in the desired state when feature is SQLENGINE 26ms (23ms|4ms)
##[error]       Expected no exception to be thrown, but an exception "System.InvalidOperationException: 'Rs' is not a valid value for setting 'FEATURES'

**Comment by johlju** [comment]
> Hi @johlju What are the available features for 2022 or how to get that? any idea?

Can be determined here: https://learn.microsoft.com/en-us/sql/database-engine/install-windows/install-sql-server-from-the-command-prompt?view=sql-server-ver16#Feature

**Comment by sara921-spec** [comment]
> Can be determined here: https://learn.microsoft.com/en-us/sql/database-engine/install-windows/install-sql-server-from-the-command-prompt?view=sql-server-ver16#Feature

hmm, RS is a valid feature for 2022

**Comment by johlju** [comment]
It says in the Description-field "Applies to: SQL Server 2016 (13.x) and earlier versions". So not part of 2022 unfortunately.

**Comment by sara921-spec** [comment]
Hi @johlju can U review this when u get a chance and let me know!

**Comment by sara921-spec** [comment]
> Hi @johlju can U review this when u get a chance and let me know!

@johlju ? Just following up, if u have any thoughts on this PR? We are looking to automate the SQL 2022 installs and this parameter is something we need to make our installs configured properly with selecting the correct license

**Comment by johlju** [comment]
I will get back to this as soon as I have time. It is on my todo list but have had less time recently to do community work due to regular work and family obligations. 😊

**Comment by dan-hughes** [comment]
@sara921-spec, please can you add an example and an integration test.

**Review by dan-hughes — COMMENTED** [review]
> Reviewed 1 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044)* status: 3 of 4 files reviewed, 7 unresolved discussions (waiting on @johlju and @sara921-spec)

___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 325 at r7](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O7-6H0y2W8LM9r9_cf_:-O7-6H0y2W8LM9r9_cfa:b-x4nsdu) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/e9c027710fab27637fe6a6efa0ffe207f4006887/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L325)):*
> ```PowerShell
>         {
>            # Grab the value of ProductCoveredBySA from the registry based on the instance
>            $getTargetResourceReturnValue.ProductCoveredBySA = Get-RegistryPropertyValue -Path "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL$($SqlVersion).$($InstanceName)\Setup" -Name "IsProductCoveredBySA"
> ```

Can the parameters of Get-RegistryPropertyValue be added to a hash table to reduce the line length.
___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 1312 at r7](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O7-721dBNzwZniWromR:-O7-721eEiywrAj-Sw8_:bkrrv0w) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/e9c027710fab27637fe6a6efa0ffe207f4006887/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L1312)):*
> ```PowerShell
>     if ($PSBoundParameters.ContainsKey('ProductCoveredBySA'))
>     {
>         if ($ProductCoveredBySA)
> ```

The additional `if` check is not required here.
___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 1314 at r7](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O7-98V-1Lhecdb4dmyK:-O7-98V-1Lhecdb4dmyL:b-j66ho5) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/e9c027710fab27637fe6a6efa0ffe207f4006887/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L1314)):*
> ```PowerShell
>         if ($ProductCoveredBySA)
>         {
>             $setupArguments['ProductCoveredBySA'] = $ProductCoveredBySA
> ```

I know it's copied from other code around it but typically this is accessed using `$setupArguments.ProductCoveredBySA`
___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 1800 at r7](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O7-9QwL7mOivTG0jxNx:-O7-9QwL7mOivTG0jxNy:b-c0xqer) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/e9c027710fab27637fe6a6efa0ffe207f4006887/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L1800)):*
> ```PowerShell
>         Default value is False.
> 
>         Not used in Test-TargetResource.    
> ```

Why is it not used? Can the instance be reinstalled/repaired/changed if the value is not in the desired state?
___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.schema.mof` line 13 at r7](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O7-9j0B7-mk5qwvO2BU:-O7-9j0B7-mk5qwvO2BV:bk5e022) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/e9c027710fab27637fe6a6efa0ffe207f4006887/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.schema.mof#L13)):*
> ```
>     [Write, Description("_SQL Server_ instance ID (if different from parameter **InstanceName**).")] String InstanceID;
>     [Write, Description("Product key for licensed installations.")] String ProductKey;
>     [Write, Description("Specifies the license coverage for SQL Server.True indicates it's covered under Software Assurance or SQL Server subscription.False, or omitting the parameter, indicates it's covered under a SQL Server license..")] Boolean ProductCoveredbySA;
> ```

The description typically only needs to describe what it does not what the values mean as this is in the parameter help. 
e.g. `Specifies the Software Assurance license coverage for SQL Server.`

**Comment by sara921-spec** [comment]
> @sara921-spec, please can you add an example and an integration test.

Does this needs a example and an integration test? As this is not a new resource and only a setting (I was comparing the existing examples and could not figure out how to add one for this yet, so just wanted to check if I make sense or it is something mandatory)

**Comment by dan-hughes** [comment]
> > @sara921-spec, please can you add an example and an integration test.
> 
> Does this needs an example and an integration test? As this is not a new resource and only a setting (I was comparing the existing examples and could not figure out how to add one for this yet, so just wanted to check if I make sense or it is something mandatory) (new here😊)

Ideally yes.

Example you can copy [this](https://github.com/dsccommunity/SqlServerDsc/blob/3c3092976409645d74f58707331f66ffe1967127/source/Examples/Resources/SqlSetup/7-InstallDefaultInstanceSingleServer2016OrLater.ps1) and add the extra parameter and update the 2016 references. It's so people know what a fully populated config looks like. 

Looking at the integration tests it might be a little tricky as there is no 2022 specific test and it uses the Eval media, which I'm not sure if the new parameter will work with. I'll defer decision to @johlju on this one.

**Review by dan-hughes — COMMENTED** [review]
Just the last few bits left and I can pass it off to someone with merge rights :-).

> Reviewed 1 of 1 files at r7, 2 of 2 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-:-O79F7Xf9Oj4rZxy3OHP:boko248)* status: all files reviewed, 5 unresolved discussions (waiting on @sara921-spec)

___
*[`source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1` line 325 at r7](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O7-6H0y2W8LM9r9_cf_:-O79EQij46E9nGFJsep8:b-kf1h5f) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/e9c027710fab27637fe6a6efa0ffe207f4006887/source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1#L325)):*
<details><summary><i>Previously, dan-hughes (Daniel Hughes) wrote…</i></summary><blockquote>

Can the parameters of Get-RegistryPropertyValue be added to a hash table to reduce the line length.
</blockquote></details>

No use splatting



_Code snippet:_
```PowerShell
$getRegistryPropertyParams = @{
   Path = "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL$($SqlVersion).$($InstanceName)\Setup"
   Name = 'IsProductCoveredBySA'
}
```
___
*[`source/Examples/Resources/SqlSetup/8-InstallDefaultInstanceSingleServer2022OrLater.ps1` line 44 at r9](https://reviewable.io/reviews/dsccommunity/SqlServerDsc/2044#-O79Esne0OsufIRBQEJR:-O79Esne0OsufIRBQEJS:b-e16g0h) ([raw file](https://github.com/dsccommunity/SqlServerDsc/blob/869a1a5e95378c7f4314219a14424fa7ce4b2e22/source/Examples/Resources/SqlSetup/8-InstallDefaultInstanceSingleServer2022OrLater.ps1#L44)):*
> ```PowerShell
>     {
>         #region Install prerequisites for SQL Server
>         WindowsFeature 'NetFramework35'
> ```

Is .NET 3.5 and 4.5 still required on supported OSes?

**Comment by sara921-spec** [comment]
> Is .NET 3.5 and 4.5 still required on supported OSes?

No, It comes by default

**Comment by johlju** [comment]
.Net 3.5 It is required for some features in some SQL Server versions that the resource probably still can install. I suggest it can be updated in another PR to determine when legacy dependencies are needed. 🤔

**Comment by johlju** [comment]
It seems there is a syntax error here: https://dev.azure.com/dsccommunity/SqlServerDsc/_build/results?buildId=9394&view=logs&jobId=5d97f74b-71f9-5a7c-9686-bc75158bbbc4&j=18b1a044-2ebe-5d7a-c679-f09fca0b67aa&t=d83cb86e-a475-59d3-8144-5a4579d10adc

**Comment by sara921-spec** [comment]
> Just the last few bits left and I can pass it off to someone with merge rights :-).

> No use splatting
> _Code snippet:_

> ```powershell
> $getRegistryPropertyParams = @{
>    Path = "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL$($SqlVersion).$($InstanceName)\Setup"
>    Name = 'IsProductCoveredBySA'
> }

@dan-hughes  Does it not needed to be passed like below? I mean for ($getTargetResourceReturnValue?)

$getTargetResourceReturnValue.ProductCoveredBySA = Get-RegistryPropertyValue @getRegistryPropertyParams

**Comment by dan-hughes** [comment]
> > Just the last few bits left and I can pass it off to someone with merge rights :-).
> 
> > No use splatting
> > _Code snippet:_
> 
> > ```powershell
> > $getRegistryPropertyParams = @{
> >    Path = "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL$($SqlVersion).$($InstanceName)\Setup"
> >    Name = 'IsProductCoveredBySA'
> > }
> > ```
> 
> @dan-hughes Does it not needed to be passed like below? I mean for ($getTargetResourceReturnValue?) or that line is not needed? Thought we need that gettargetrespurcereturnvalue
> 
> $getTargetResourceReturnValue.ProductCoveredBySA = Get-RegistryPropertyValue @getRegistryPropertyParams

You are correct that is how you pass a splat to a command. The last commit you made looks to be good.

**Comment by dan-hughes** [comment]
@sara921-spec, run a format on the module file and commit again, it should fix the HQRM test.

**Comment by sara921-spec** [comment]
@dan-hughes  It actually looked for single quotes as mentioned here https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#correct-format-for-hashtables-or-objects)., 

I believe we are looking good now! do u see any other updates required?

**Comment by dan-hughes** [comment]
> @dan-hughes It actually looked for single quotes as mentioned here https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#correct-format-for-hashtables-or-objects).,
> 
> I believe we are looking good now! do u see any other updates required?

The path string was not the issue, it was the formatting. I've just cloned your repo and opened it in VSCode.

image

If I put your double quotes back so the string has the correct values in and format it it looks like this.

image

**Comment by sara921-spec** [comment]
@dan-hughes interesting, it didn't show up for me, do u want the double quotes back? because from the doc and there was a line of code in the same file single quotes was used for splatting like below on line no "2764" As I tried few other things before I updated it to single quotes, only then I got the successful build

    $getRegistryPropertyValueParameters = @{
        Path = 'HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\Instance Names\SQL'
        Name = $InstanceName
    }

**Comment by sara921-spec** [comment]
@dan-hughes does it looks good nw?

**Comment by dan-hughes** [comment]
@sara921-spec, I'll look at this over the weekend.

File: source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1 [hunk 1]
```diff
@@ -518,6 +529,11 @@ function Get-TargetResource
     .PARAMETER ProductKey
         Product key for licensed installations.
 
+   .PARAMETER PRODUCTCOVEREDBYSA
```
Comment (dan-hughes) [inline source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1]: Please set the correct casing of the parameter name. Also need an extra space before `.PARAMETER`

File: source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1 [hunk 2]
```diff
@@ -318,6 +319,16 @@ function Get-TargetResource
             $getTargetResourceReturnValue.SqlTempdbLogFileGrowth = $currentTempDbProperties.SqlTempdbLogFileGrowth
         }
 
+        if ($sqlVersion -ge 16)
+        {
+           # Grab the value of ProductCoveredBySA from the registry based on the instance
+           $getRegistryPropertyParams = @{
+               Path = "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL$($SqlVersion).$($InstanceName)\Setup"
+               Name = 'IsProductCoveredBySA'
```
Comment (dan-hughes) [inline source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1]: Only 3 spaces indented.

File: source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1 [hunk 3]
```diff
@@ -318,6 +319,16 @@ function Get-TargetResource
...
+        }
+           $getTargetResourceReturnValue.ProductCoveredBySA = Get-RegistryPropertyValue @getRegistryPropertyParams
```
Comment (dan-hughes) [inline source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1]: Same as above.

File: source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1 [hunk 4]
```diff
@@ -1767,6 +1793,13 @@ function Set-TargetResource
     .PARAMETER ProductKey
         Product key for licensed installations.
 
+    .PARAMETER PRODUCTCOVEREDBYSA
```
Comment (dan-hughes) [inline source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1]: Parameter name casing same as `Set-TargetResource`

File: source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1 [hunk 5]
```diff
@@ -1767,6 +1793,13 @@ function Set-TargetResource
...
+        Specifies the license coverage for SQL Server. True indicates it's covered under Software Assurance or SQL Server subscription.
+        False, or omitting the parameter, indicates it's covered under a SQL Server license.
+        Default value is False.
+
+        Not used in Test-TargetResource.
```
Comment (dan-hughes) [inline source/DSCResources/DSC_SqlSetup/DSC_SqlSetup.psm1:1801]: Why is it not checked in `Test-TargetResource`? If the resource is not set correctly should the setup not rerun with the correct parameters?

File: tests/Unit/DSC_SqlSetup.Tests.ps1 [hunk 6]
```diff
@@ -828,6 +831,18 @@ Describe 'SqlSetup\Get-TargetResource' -Tag 'Get' {
                     $result.Features | Should -Match 'BOL\b'
                     $result.Features | Should -Match 'MDS\b'
                 }
+                if ($MockSqlMajorVersion -in ('16'))
```
Comment (dan-hughes) [inline tests/Unit/DSC_SqlSetup.Tests.ps1]: Should be `elseif`.

File: tests/Unit/DSC_SqlSetup.Tests.ps1 [hunk 7]
```diff
@@ -972,6 +987,18 @@ Describe 'SqlSetup\Get-TargetResource' -Tag 'Get' {
                     $result.Features | Should -Match 'BC\b'
                     $result.Features | Should -Match 'SDK\b'
                 }
+                if ($MockSqlMajorVersion -in ('16'))
```
Comment (dan-hughes) [inline tests/Unit/DSC_SqlSetup.Tests.ps1]: Should be `elseif`

File: tests/Unit/DSC_SqlSetup.Tests.ps1 [hunk 8]
```diff
@@ -1205,6 +1232,18 @@ Describe 'SqlSetup\Get-TargetResource' -Tag 'Get' {
...
```
Comment (dan-hughes) [inline tests/Unit/DSC_SqlSetup.Tests.ps1]: Should be `elseif`

File: tests/Unit/DSC_SqlSetup.Tests.ps1 [hunk 9]
```diff
@@ -1514,6 +1553,18 @@ Describe 'SqlSetup\Get-TargetResource' -Tag 'Get' {
...
```
Comment (dan-hughes) [inline tests/Unit/DSC_SqlSetup.Tests.ps1]: Should be `elseif`

**Review by dan-hughes — CHANGES_REQUESTED** [review]
Unit Tests:
ContextBlock 'When SQL Server version is <MockSqlMajorVersion> and the system is not in the desired state for default instance'

- It needs to add the check that the value is false.

**Comment by sara921-spec** [comment]
> Unit Tests: ContextBlock 'When SQL Server version is and the system is not in the desired state for default instance'
> 
> * It needs to add the check that the value is false.

@dan-hughes  False for the parameter productcoveredbysa? In the past johan had the below thought (or am I not reading your request correctly?)

image

**Comment by dan-hughes** [comment]
> > Unit Tests: ContextBlock 'When SQL Server version is and the system is not in the desired state for default instance'
> > 
> > * It needs to add the check that the value is false.
> 
> @dan-hughes False for the parameter productcoveredbysa? In the past johan had the below thought (or am I not reading your request correctly?)
> 
> image

Apologies for not being clear, 

It was more of adding the `$result.ProductCoveredBySA | Should -BeFalse` check [here](https://github.com/dsccommunity/SqlServerDsc/blob/3c3092976409645d74f58707331f66ffe1967127/tests/Unit/DSC_SqlSetup.Tests.ps1#L379-L405) for the as the resource should return the default value for the type.

It looks like you have resolved all the other issues so once this is done and it builds it's good to go.

**Comment by dan-hughes** [comment]
I'm not sure where I got that Context block title from though as I've just searched and cannot find it!

File: tests/Unit/DSC_SqlSetup.Tests.ps1 [hunk 10]
```diff
@@ -2707,6 +2823,52 @@ Describe 'SqlSetup\Set-TargetResource' -Tag 'Set' {
             }
         }
 
+        Context 'When installing the database engine and ProductcoveredBySA is true' {
```
Comment (dan-hughes) [inline tests/Unit/DSC_SqlSetup.Tests.ps1:2784]: I'm not in front of a big enough screen to look at the whole test file, but this test looks to be a copy + paste of the test added at line 2484. Is this supposed to be here?

Comment (sara921-spec) [reply tests/Unit/DSC_SqlSetup.Tests.ps1:2784]: ah, line 2826 is the actual test, I don't know how 2484 got there, I should have misplaced it somehow. let me remove it

Comment (sara921-spec) [reply tests/Unit/DSC_SqlSetup.Tests.ps1:2784]: @dan-hughes all set I guess -:)

**Review by dan-hughes — APPROVED** [review]

**Comment by johlju** [comment]
@sara921-spec awesome work on this, thank you! Also big thanks to @dan-hughes for the review! 🙇‍♂️

**Pull request merged**